### ISSUE TITLE
Design Idea: Multiplanetary

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#179]
This thread is a follow-on discussion of the [comment](https://github.com/mars-sim/mars-sim/issues/130#issuecomment-455390007) that we should anticipate the contribution of `Earth` and `Moon` to the development of the human settlement of Mars.

In the foreseeable future, it is reasonably expected that the political clout of some Earth's governments and/or multi-national corporations may heavily influence the developing Martian colonies in various ways.

Until the day when the Martian settlements are able to build its own industrial base that can loosen the reliance of Earth for imports of goods, Mars will be dependent upon Earth, just like how old colonial powers would exercise political, social and economical controls over established colonies on another continent.

Let's assume that permanent lunar bases will also have been established by the time we landed on Mars. If the lunar bases are able to export any industrial resources by then, provided that a space transportation system is in place, Mars could as well be importing resources excavated or produced from the Moon.

Feel free to write how you envision this tri-system relationship would work in a macroscopic scale and what we may model in mars-sim specifically.

**Comment by Urwumpe** [comment]
Well, I think the most likely goods exchanged between Moon and Mars would be immaterial goods. Information or services. Maybe people eventually, like people from a Moon expedition going to Mars.

Which organisational actors do we want to include there? A static set of transnat companies or do we generate them during simulation start or even during runtime?

**Comment by mokun** [comment]
> Well, I think the most likely goods exchanged between Moon and Mars would be immaterial goods. Information or services. Maybe people eventually, like people from a Moon expedition going to Mars.

Sure you spell out 3 things : 
1. Goods
2. Information
3. People

Let's discussion each of them and first out some conceptual models on how we approach the implementation.

**Comment by mokun** [comment]
For (1) the Goods Exchange, I'd like to work out a credit system. 

Currently, the Value Point (VP) in mars-sim is *one-dimensional* because it measures the possible demand by 
(A). Giving an estimate on based on the projected usage 
(B). Comparing between how much it **once** had in all inventories and what it has **now**. 

The higher the VP, the more demanding the amount resource of interest.

![vp1](https://user-images.githubusercontent.com/1168584/51559467-cee9b200-1e36-11e9-8296-3ae403ff4061.jpg)

Under this system, usually all raw materials would normally be having very high VPs.

We would like to convert this system into a monetary system so that fair trading on resources would be possible between Earth, Moon and Mars.

**Comment by mokun** [comment]
Q: what are some other prerequisites for doing fair trading on resources between Earth, Moon and Mars ?

**Comment by Urwumpe** [comment]
What do you call fair?

If you assume a free market situation, its only fair if both supply and demand are balanced. If you have far more supply than demand for example, the customers can dictate the price. 

If the only entity delivering goods to Mars is a single agency or company, it could dictate the transport prices to Mars.

During the colonization phase that we can simulate right now, having a free market situation for all goods would be deadly. At least food, water and other life support/existencial goods should be excluded from a free market and have fixed prices or be traded by other means.  For example, it would make no sense for Mars to export its food to Earth (And of course, refer to "Moon is a Harsh Mistress" there as well) 

But for example: We sure can exclude that we can create nuclear fuel on Mars for a while. Luckily the reactors don't consume their fuel elements right now. But if they would, it would be inevitable that Earth has to supply them. Using a free market would fail, since there is only one supplier. Also, the settlers need them for producing power. The only way this could work would be a contract that Earth supplied the needed fuel for the nuclear reactors until the EOL.

Of course, Mars could also rent the reactor from Earth... since Mars also has other power sources available (but less powerful) this would still be a oligopoly, but much less than when trading fuel elements. Mars would then need to reduce the need for nuclear reactors. 

With moon, we could also trade Helium-3 with it to fuel fusion reactors. But those reactors are still not working well on Earth and Helium-3 is also rare on the moon. And transporting the Helium-3 to Earth would be cheaper, but launching Helium-3 from the moon to Mars would be cheaper than transporting nuclear fuel elements from Earth to Mars.

**Comment by mokun** [comment]
Okay now we have 4 things to consider when it comes to trading : 
1. Goods
2. Information
3. People
4. Power

**Comment by mokun** [comment]
> During the colonization phase that we can simulate right now, having a free market situation for all goods would be deadly. At least food, water and other life support/existencial goods should be excluded from a free market and have fixed prices or be traded by other means. 

Yes they said there's no such thing as a *Free Market* in any civilized worlds because there's always pricing fixing and tariff and other measures for trade protection. 

Yes we want to promote price stability and long term investment rather than short term trading. It's unthinkable to allow the price of a raw material to go free fall. It will hurt the mining industry or production. 

We will need to introduce control measures to dampen any sudden price hike and smooth the increase/decrease.

**Comment by mokun** [comment]
BTW, I started watching the first 3 episode of the *Nat Geo's : Mars Season 2* a couple months ago. When it comes to developing Mars, they depicted the scenario or contention between the altruistic goals of a Mars consortium and key drivers of a private company. They don't speak the same language and would never understand each other at all.

**Comment by Urwumpe** [comment]
> How about we create a set of _economic policies_ check boxes ?
> 
> When a settlement arrives at a certain stage of development, a policy may be unlocked ?
> 

I like that idea, but I would not make it depend on stage of development. I would maybe even remove direct control over it in commander mode. 

What about having some "Constitution of Mars" or "Settlement Charta" there, that defines how things are run? Like having political entities and assign powers to them. Or define how resources should be allocated? And make the decisions of the settlers depend on personal motivations and social relations?

For example, settlers could discuss a certain policy with others. Or petition somebody to decide in a certain way. 

And of course, in the beginning, the organisation sponsoring a settlement might have all power there, so everything is decided on Earth. Maybe there should be something like allowing martian or transnat shadow governments there that have their own set of policies, which non-members might not be aware of. And informers inside the groups might tell other groups about the policies and goals of another group... 

But in sandbox game the UI with checkboxes makes sense, also for doing tests. Maybe it should be an option for it.

**Comment by mokun** [comment]
So  based on what we've discussed and shared here and there, we can think of two categories of policies, namely, *Economic Policies* and *Governmental Policies* that we can implement.

The *Economic Policies* have to do with moderating/alternating the way goods and services are produced, exchanged and consumed by settlers and settlements.

The *Governmental Policies" no doubt will impact and shape how the society develops and structures.

**Comment by mokun** [comment]
One question is : what *population threshold* is it supposed to reach in order to activate these economic/political policies ?

In terms of hierarchy, in ChainOfCommand, we define the need of a commander, sub-commander, mayor as follows : 

```
	public static final int POPULATION_WITH_SUB_COMMANDER = 12;
	public static final int POPULATION_WITH_MAYOR = 48;
	public static final int POPULATION_WITH_COMMANDER = 4;
```

**Comment by mokun** [comment]
> using the ERG model in the background for determining how much thought settlers put to certain policies?


ERG is interesting. I only heard of Maslow's and it certainly needs some improvement because human beings are way more complicated and our needs can be parallel and interconnected and not necessarily linear or hierarchical.

> Some hardwiring would of course still needed somewhere - but I think, it could be kept at a minimum. We need starting parameters for the simulation, but we don't need to define them in code - the initial settlement organisation charts could be defined in JSON and the population limit just get a hard minimum by this chart. 

True. Let's start moving these hardwiring code to JSON. Yes we can specify how each space agency having their own sets of rules in JSON for selecting candidates/settlers over different sets of skills, etc.

**Comment by mokun** [comment]
Also, we can move the criteria for job and role assignment to JSON.

Recently, I was looking into how previous developers did the job demand and prospect calculation for determining who takes up what job. 

And yesterday I was tweaking how roles are assigned. Sometimes rigid rules don't work well but that's because our algorithms have holes. 

Of course, human beings can offer better judgment to pick and choose. Perhaps we should let players choose more. 

All we have to do is to write a UI to let them choose at the start of the simulation. The algorithm will present them the result but players should have the final call to choose many more things.

On the other hand, we do want smart A.I. and aid players so that they don't have to micromanage the settlement.

**Comment by mokun** [comment]
At the start of the sim, some goods have incredibly low 'Value Point (VP)`.

They may be displayed as having the value of `0.00` when the true value could be, say, `0.0031`. 

The last two digits get cut off because we set the table to be able to handle only up to the last 2 decimal places. 

> Note: normally having too many digits work against us human and almost doesn't register to our senses with the all those extra digits.

I'll need to hack it in such a way that if an EVA suit has a VP of `375.0235`, it will be truncated to just `375.02` for readability sack.

But for hydrogen, it might have a value of `0.0001234`, it will still show the value of `0.00012` (at least showing two digits at the end.

**Comment by mokun** [comment]
The plain old `DecimalFormat ` class in java can only handle *static* decimal places. For instance,

```
	private DecimalFormat dec0Formatter = new DecimalFormat("#,###,###");
	private DecimalFormat dec1Formatter = new DecimalFormat("#,###,##0.0");
	private DecimalFormat dec2Formatter = new DecimalFormat("#,###,##0.00");

```
Goods with large VPs are okay.

![large VPs](https://user-images.githubusercontent.com/1168584/66620099-26ca7380-eb94-11e9-81cf-312e23d1bc20.jpg)

But goods with low VPs are bad : 

![VP less than 0 00](https://user-images.githubusercontent.com/1168584/66620122-3944ad00-eb94-11e9-8b73-0516892d9030.jpg)

**Comment by Urwumpe** [comment]
How about adding a batch size to the goods, so that instead of buying one good with a very low VP, you instead have to order a larger batch, eg. instead of 0.001 per kg wheat, you order 1000 kg wheat for 1 VP

**Comment by mokun** [comment]
@Urwumpe 

Question: did you witness any `unreasonable` trading between 2 settlements involving minuscule quantity/amount of resources ?

Good points in trading for batch size or whatever volume each container can hold :D 

Also, I'm wondering if the history of each trade was recorded so that we can look back ? 

In the past, I saw the stacktrace would print what were being exchanged.

I can see the Credit tab on a settlement that showing the VP exchange between settlements but not the content of a trade.

**Comment by Urwumpe** [comment]
I did see some weird trades, but mostly annoying had been the long loading/unloading of rovers there.  I needed to look at the mission view for knowing what was traded. 

Some trade statistics would be great, much better might be supporting accounting. Like having an internal audit AI for the settlements that keeps track of the trades and production, and a balance sheet tab for a settlement that allows viewing the results. 

One tiny scientific problem there might be, that we have no currency, not even a virtual one (VPs are different). I don't know how accounting would work for barter trades, but I am sure, somebody else did already think about that (maybe 3000 years ago). Of course, another option might be introducing a virtual currency unit to make the accounting simpler. Or use the unit "hour of work" there.

**Comment by mokun** [comment]
> One tiny scientific problem there might be, that we have no currency, not even a virtual one (VPs are different). 

True. It's inevitable that some form of currency will arise to ease the trade between settlements

So let's be intentional in discussing how we might create a monetary system :D  

So far we have : 

`VP = demand / supply * modifier`

> Or use the unit "hour of work" there.

Yes. we can set it up this way,

`M$ = supply / demand * labor cost * tech skill * power cost * facility cost * modifier`

The *tech skill* are simply the skill level of the settlers who manufacture the item/equipment

The *labor cost* is how long they work on it.

The *power cost* is the electricity/heat needed in the production.

We'll need to come up with the *facility cost*. May be it the lifetime cost of the building.

The take-away is that since each settlement will compute her own M$ for each resource/part/equipment, as each settlement will follow a different *developmental path*.

As a result, the M$ for, say, an EVA suit, in colony A may end up lower than the one from colony B consistently.

This way, it creates more incentives for settlements to trade with others to maximize their resource pool.

**Comment by Urwumpe** [comment]
What about separating price, cost and value? Those are different things, and tracking them separatly could allow better control of the economy. Also, this would not automatically require us to use a currency behind. We could also apply this to bartering transactions. 

Your model does not include materials and consumables, which I think could be a problem for proper cost calculations and thus pricing.

What I also wonder is (maybe we can test it in MSP as optional feature), if a "life support unit" could be used to represent the life support resources consumed per sol and per settler and which could be included into the cost model for each good to simplify the calculations. 

A good that takes a lot of time and settlers to produce would be more costly in general, but would also be more costly for a settlement that has only limited life support units available or has a more expensive life support system (for example, no greenhouse).

**Comment by mokun** [comment]
> Your model does not include materials and consumables, which I think could be a problem for proper cost calculations and thus pricing.

I believe for a given `Part`, say a resistor, since it's created by combing a few of the `Amount Resource`, the VP of the it is the summation of the VP of each of them.

```
		<!-- 2017-09-11 Added manufacture resisters -->
	<process name="manufacture resistors" tech="1" skill="2" work-time="50" process-time="200"
			power-required="1">
		<inputs>
			<resource name="aluminum" amount="0.1" />
			<resource name="aluminum oxide" amount="0.1" />
			<resource name="zinc" amount="0.1" />
			<resource name="nickel" amount="0.1" />					
			<resource name="iron" amount="0.1" />	
			<resource name="carbon" amount="1" />		
		</inputs>
		<outputs>
			<part name="resistor" number="400" />
		</outputs>
	</process>
```

So let me modify the formula   : 

The price of a resister is 

`M$ = Summation of (S_i / D_i * L_i * TS_i * P_i * F_i * M_i), for i is 1 to 6`

**Comment by mokun** [comment]
Or 

`M$ = Summation of ( M_i *  (S_i / D_i  + L_i + TS_i + P_i + F_i  + O_i + C_i ) )`

O_i is the operating cost and C_i is the cost of the consumables

What can we right now define as consumables ?

**Comment by mokun** [comment]
> What about separating price, cost and value? Those are different things, and tracking them separatly could allow better control of the economy. 

Yes we could. 

And I'll make each Good object record its previous cost, value and price, instead of what's currently done using `GoodManager`.

**Comment by mokun** [comment]
> Also, this would not automatically require us to use a currency behind. We could also apply this to bartering transactions.

Sure. we can **black market** in future that allows some *runaway* settlers to barter or do *swapmeet*. 

While their transactions will not be noticeable and registered by central authority, they should still somewhat affect the economy. It depends on what we discuss how it can be implemented in future.

**Comment by mokun** [comment]
>  so that instead of buying one good with a very low VP, you instead have to order a larger batch, 

https://github.com/mars-sim/mars-sim/blob/96b3fcab7504691c53bed9b67ec527d51ee66067/mars-sim-core/src/main/java/org/mars_sim/msp/core/person/ai/mission/TradeUtil.java#L802-L826

I hope this method in `TradeUtil` would take care of trading the resources according to the max capacity of a container that holds the ~Amount Resource` that is being traded.

**Comment by mokun** [comment]
@Urwumpe ,

> I did see some weird trades, but mostly annoying had been the long loading/unloading of rovers there. I needed to look at the mission view for knowing what was traded.

I wonder why it takes so long to load and unload.

> Some trade statistics would be great, much better might be supporting accounting. Like having an internal audit AI for the settlements that keeps track of the trades and production, and a balance sheet tab for a settlement that allows viewing the results.

Yes I'm planning to add the stats showing the history of each transaction.

And I'll be adding a few *administrative* policies on starting a mission. 

Say the commander may choose if he allows settlers to start a *Trade Mission* or not. He may only choose if his settlement would accept other settlements to send a *Trade Mission* over.

May be we can even set some preconditions or criteria for every trade negotiation.

**Comment by mokun** [comment]
> Some trade statistics would be great, much better might be supporting accounting. Like having an internal audit AI for the settlements that keeps track of the trades and production, and a balance sheet tab for a settlement that allows viewing the results.

I just looked at `Trade` and `TradeUtil`.

https://github.com/mars-sim/mars-sim/blob/592faf0b5f6dc9500e16b06eb15ceb5cb4da400a/mars-sim-core/src/main/java/org/mars_sim/msp/core/person/ai/mission/Trade.java#L76-L78

So, it turns out `Trade` has a Map that records the profit of each trade with an inner class called `TradeProfitInfo` that's inside `Trade`.

See below : 
https://github.com/mars-sim/mars-sim/blob/592faf0b5f6dc9500e16b06eb15ceb5cb4da400a/mars-sim-core/src/main/java/org/mars_sim/msp/core/person/ai/mission/Trade.java#L1075-L1087

We could expand that to cover the exact goods involved in each trade.

For now, the buying and selling goods are spread over 3 other separate Maps as follows in `Trade` : 

https://github.com/mars-sim/mars-sim/blob/592faf0b5f6dc9500e16b06eb15ceb5cb4da400a/mars-sim-core/src/main/java/org/mars_sim/msp/core/person/ai/mission/Trade.java#L93-L95

**Comment by mokun** [comment]
Closing this long thread. Price/Cost/Value of a good are being worked on lately in https://github.com/mars-sim/mars-sim/issues/598 and https://github.com/mars-sim/mars-sim/issues/591.
